qiime2/plugins.py

```diff
@@ -78,15 +82,23 @@ def render(self):
     def get_example_data(self):
         return {r: f() for r, f in self._init_data_refs.items()}
 
-    def _template_action(self, action_f, input_opts, output_opts):
-        output_opts = list(output_opts.values())
+    def _template_action(self, action_f, inputs, params, mds, outputs):
+        output_opts = [x for x, _ in outputs.values()]
         if len(output_opts) == 1:
             output_opts.append('')
         output_vars = ', '.join(output_opts)
 
         t = '%s = %s(\n' % (output_vars.strip(), action_f.id)
-        for k, v in input_opts.items():
+
+        for k, (v, _) in inputs.items():
+            t += '    %s=%s,\n' % (k, v)
+
+        for k, (v, _) in params.items():
+            t += '    %s=%r,\n' % (k, v)
```

Comment:
This is the bulk of the bug-fix, changing from a str to a repr when templating vanilla parameters.

Reply:
I think the changes introduced here resulted in slightly different ordering, revealed by [this test failing](https://github.com/thermokarst-forks/qiime2/blob/6930d5dbd08f0533367f3ed72481de7e024aa9e4/qiime2/tests/test_artifact_api.py#L284-L289):.

```
AssertionError: 'from[362 chars]ts=ints,\n    num1=3,\n    optional1=ints_b,\n    num2=4,\n)\n' != 'from[362 chars]ts=ints,\n    optional1=ints_b,\n    num1=3,\n    num2=4,\n)\n'
  from qiime2.plugins.dummy_plugin.methods import optional_artifacts_method

  output, = optional_artifacts_method(
      ints=ints,
      num1=1,
  )

  output, = optional_artifacts_method(
      ints=ints,
      num1=1,
      num2=2,
  )

  ints_b, = optional_artifacts_method(
      ints=ints,
      num1=1,
      num2=None,
  )

  output, = optional_artifacts_method(
      ints=ints,
+     optional1=ints_b,

      num1=3,
-     optional1=ints_b,
      num2=4,
  )
```

I think it might be fine to adjust the expected value of that test, i.e.


```diff
modified   qiime2/tests/test_artifact_api.py
@@ -283,8 +283,8 @@ ints_b, = optional_artifacts_method(
 
 output, = optional_artifacts_method(
     ints=ints,
-    num1=3,
     optional1=ints_b,
+    num1=3,
     num2=4,
 )
 """

```

Reply:
I think the changes introduced here resulted in slightly different ordering, revealed by [this test failing](https://github.com/thermokarst-forks/qiime2/blob/6930d5dbd08f0533367f3ed72481de7e024aa9e4/qiime2/tests/test_artifact_api.py#L284-L289):.

```
AssertionError: 'from[362 chars]ts=ints,\n    num1=3,\n    optional1=ints_b,\n    num2=4,\n)\n' != 'from[362 chars]ts=ints,\n    optional1=ints_b,\n    num1=3,\n    num2=4,\n)\n'
  from qiime2.plugins.dummy_plugin.methods import optional_artifacts_method

  output, = optional_artifacts_method(
      ints=ints,
      num1=1,
  )

  output, = optional_artifacts_method(
      ints=ints,
      num1=1,
      num2=2,
  )

  ints_b, = optional_artifacts_method(
      ints=ints,
      num1=1,
      num2=None,
  )

  output, = optional_artifacts_method(
      ints=ints,
+     optional1=ints_b,

      num1=3,
-     optional1=ints_b,
      num2=4,
  )
```

I think it might be fine to adjust the expected value of that test, i.e.


```diff
modified   qiime2/tests/test_artifact_api.py
@@ -283,8 +283,8 @@ ints_b, = optional_artifacts_method(
 
 output, = optional_artifacts_method(
     ints=ints,
-    num1=3,
     optional1=ints_b,
+    num1=3,
     num2=4,
 )
 """

```

Reply:
Thanks! I'm not quite there yet, still thinking about approaches for this PR, so nothing is set in stone, but noted!

Reply:
Forgot this was a draft, but I just happened to stumble upon this today so thought I'd share 🥨